### PR TITLE
Display skyBox

### PR DESCRIPTION
**Pull request by Nokitoo**

### CODE REVIEW

File: include/lug/Graphics/Builder/SkyBox.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+class Renderer;
+
+namespace Builder {
+
+class LUG_GRAPHICS_API SkyBox {
+    friend Resource::SharedPtr<lug::Graphics::Render::SkyBox> lug::Graphics::Vulkan::Builder::SkyBox::build(const ::lug::Graphics::Builder::SkyBox&);
+
+public:
+    enum class Face: uint8_t {
+        POSITIVE_X = 0,
```
Comment (saveman71) [inline include/lug/Graphics/Builder/SkyBox.hpp]: `PositiveX` https://lugdunum3d.github.io/doc/guidelines.html#enumerator-names

File: include/lug/Graphics/Builder/SkyBox.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    Render::Texture::Filter _magFilter{Render::Texture::Filter::Nearest};
+    Render::Texture::Filter _minFilter{Render::Texture::Filter::Nearest};
+    Render::Texture::Filter _mipMapFilter{Render::Texture::Filter::Nearest};
+
+    Render::Texture::WrappingMode _wrapS{Render::Texture::WrappingMode::ClampToEdge};
+    Render::Texture::WrappingMode _wrapT{Render::Texture::WrappingMode::ClampToEdge};
+
+    // Contains alls files for faces
+    std::unordered_map<Face, std::string> _faces;
```
Comment (saveman71) [inline include/lug/Graphics/Builder/SkyBox.hpp]: Transform to std::array instead

File: include/lug/Graphics/Builder/Texture.hpp [hunk 3]
```diff
@@ -16,6 +17,17 @@ namespace Builder {
...
 
+    struct Layer {
+        // TODO(nokitoo): add other infos(layers count, aspect mask, mip level, etc...)
+        std::string filename;
+    };
+
+public:
+    enum class Type {
+        Texture2D,
+        Cube
```
Comment (saveman71) [inline include/lug/Graphics/Builder/Texture.hpp]: CubeMap

File: include/lug/Graphics/Resource.hpp [hunk 4]
```diff
@@ -29,7 +29,8 @@ class LUG_GRAPHICS_API Resource {
         Light,      ///< A light
         Texture,    ///< A texture
         Pipeline,   ///< A graphical pipeline
-        Camera      ///< A camera
+        Camera,      ///< A camera
```
Comment (saveman71) [inline include/lug/Graphics/Resource.hpp]: Comment indent

File: include/lug/Graphics/Render/SkyBox.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    SkyBox(const SkyBox&) = delete;
+    SkyBox(SkyBox&&) = delete;
+
+    SkyBox& operator=(const SkyBox&) = delete;
+    SkyBox& operator=(SkyBox&&) = delete;
+
+    virtual ~SkyBox() = default;
+
+    const lug::Graphics::Resource::SharedPtr<lug::Graphics::Render::Texture> getTexture() const;
```
Comment (saveman71) [inline include/lug/Graphics/Render/SkyBox.hpp]: Resource::SharedPtr

File: include/lug/Graphics/Vulkan/Render/SkyBox.hpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    static const API::GraphicsPipeline& getPipeline();
+    static const lug::Graphics::Resource::SharedPtr<lug::Graphics::Render::Mesh> getMesh();
+
+private:
+    static bool initPipeline(Renderer& renderer);
+    static bool initMesh(Renderer& renderer);
+
+private:
+    /**
+     * @brief      Constructs a SkyBox
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/Render/SkyBox.hpp]: Fail

File: include/lug/Graphics/Vulkan/Render/SkyBox.hpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+     * @param[in]  name  The name of the SkyBox
+     * @param[in]  type  The type of the SkyBox
+     */
+    SkyBox(const std::string& name);
+
+private:
+    static API::GraphicsPipeline _pipeline;
+    static lug::Graphics::Resource::SharedPtr<lug::Graphics::Render::Mesh> _mesh;
+
+    // Store the number of skybox created to know when
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/Render/SkyBox.hpp]: Doxygen comments please

File: resources/shaders/skybox.frag [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#version 450
+
+layout (set = 1, binding = 0) uniform samplerCube skyBox;
+
+layout (location = 0) in vec3 inPos;
+
+layout (location = 0) out vec4 outColor;
+
+void main()
```
Comment (saveman71) [inline resources/shaders/skybox.frag]: () {

File: resources/shaders/skybox.vert [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+layout(std140, set = 0, binding = 0) uniform cameraBlock {
+    mat4 view;
+    mat4 proj;
+} camera;
+
+layout (location = 0) in vec3 inPos;
+
+layout (location = 0) out vec3 outPos;
+
+void main()
```
Comment (saveman71) [inline resources/shaders/skybox.vert]: () {

File: src/lug/Graphics/Vulkan/Render/Technique/Forward.cpp [hunk 10]
```diff
@@ -165,6 +169,61 @@ bool Forward::render(
...
+                    /* skyBoxBind.pipelineLayout     */ *SkyBox::getPipeline().getLayout(),
+                    /* skyBoxBind.pipelineBindPoint  */ VK_PIPELINE_BIND_POINT_GRAPHICS,
+                    /* skyBoxBind.firstSet           */ 1,
+                    /* skyBoxBind.descriptorSets     */ {&frameData.skyBoxDescriptorSet->getDescriptorSet()},
+                    /* skyBoxBind.dynamicOffsets     */ {},
+                };
+
+                frameData.renderCmdBuffer.bindDescriptorSets(skyBoxBind);
+            }
+
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Render/Technique/Forward.cpp]: Too many blank lines

File: src/lug/Graphics/Vulkan/Render/SkyBox.cpp [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        VkResult result{VK_SUCCESS};
+        API::RenderPass renderPass;
+        if (!renderPassBuilder.build(renderPass, &result)) {
+            LUG_LOG.error("SkyBox::initPipeline: Can't create render pass: {}", result);
+            return false;
+        }
+
+        graphicsPipelineBuilder.setRenderPass(std::move(renderPass), 0);
+    }
+
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Render/SkyBox.cpp]: Too many blank lines

File: src/lug/Graphics/Vulkan/Render/SkyBox.cpp [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+                /* binding.descriptorType */ VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+                /* binding.descriptorCount */ 1,
+                /* binding.stageFlags */ VK_SHADER_STAGE_FRAGMENT_BIT,
+                /* binding.pImmutableSamplers */ nullptr
+            };
+
+            descriptorSetLayoutBuilder.setBindings({binding});
+            VkResult result{VK_SUCCESS};
+            if (!descriptorSetLayoutBuilder.build(descriptorSetLayouts[1], &result)) {
+                LUG_LOG.error("SkyBox::initPipeline: Can't create pipeline descriptor: {}", result);
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Render/SkyBox.cpp]: "Can't create pipeline descriptor sets layout 1: {}"

File: src/lug/Graphics/Vulkan/Render/SkyBox.cpp [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+SkyBox::SkyBox(const std::string& name) : ::lug::Graphics::Render::SkyBox(name) {}
+
+SkyBox::~SkyBox() {
+    destroy();
+}
+
+void SkyBox::destroy() {
+    --_skyBoxCount;
+    if (!_skyBoxCount) {
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Render/SkyBox.cpp]: I prefer `_skyBoxCount == 0`

**Review by saveman71 — CHANGES_REQUESTED** [review]

**Review by Quentin01 — DISMISSED** [review]

**Review by saveman71 — DISMISSED** [review]

**Pull request merged**